File: smoke_tests/typescript/zip_slip.ts
// Zip Slip Test Cases

import * as fs from 'fs';
import * as path from 'path';

// Test 1: Extracting zip without path validation
function extractZipEntry(entry: any, destDir: string): void {
    const fileName = entry.fileName;
    const destPath = path.join(destDir, fileName);
    // VULNERABLE: No validation - fileName could be "../../../etc/passwd"
    fs.writeFileSync(destPath, entry.getData());
}

// Test 2: Tar extraction without sanitization
function extractTarFile(entryPath: string, outputDir: string, data: Buffer): void {
    // VULNERABLE: Direct path join without sanitization
    const outputPath = outputDir + '/' + entryPath;
    fs.writeFileSync(outputPath, data);
}

// Test 3: Archive extraction with normalize but no validation
function extractArchive(entry: any, baseDir: string): void {
    const normalizedPath = path.normalize(entry.name);
    const fullPath = path.join(baseDir, normalizedPath);
    // VULNERABLE: normalize doesn't prevent path traversal
    fs.mkdirSync(path.dirname(fullPath), { recursive: true });
    fs.writeFileSync(fullPath, entry.data);
}

// Test 4: Using user-provided filename
function saveUploadedArchiveEntry(filename: string, content: Buffer, uploadDir: string): void {
    // VULNERABLE: filename from archive could contain ../
    const filePath = `${uploadDir}/${filename}`;
    fs.writeFileSync(filePath, content);
}

// Test 5: Concatenating paths without validation
function unpackFile(archiveEntry: any, destination: string): void {
    const targetPath = destination + path.sep + archiveEntry.path;
    // VULNERABLE: No check if targetPath escapes destination
    const dir = path.dirname(targetPath);
    if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
    }
    fs.writeFileSync(targetPath, archiveEntry.content);
}

File: smoke_tests/rust/zip_slip.rs
// Zip Slip Test Cases

use std::fs::File;
use std::io::Write;
use std::path::Path;

// Test 1: Extracting zip without path validation
fn extract_zip_entry(entry_path: &str, dest_dir: &Path, data: &[u8]) -> std::io::Result<()> {
    // VULNERABLE: No validation - entry_path could be "../../../etc/passwd"
    let dest_path = dest_dir.join(entry_path);
    let mut file = File::create(dest_path)?;
    file.write_all(data)?;
    Ok(())
}

// Test 2: Tar extraction without sanitization
fn extract_tar_file(entry_path: &str, output_dir: &Path, data: &[u8]) -> std::io::Result<()> {
    // VULNERABLE: Direct path join without sanitization
    let output_path = output_dir.join(entry_path);
    std::fs::create_dir_all(output_path.parent().unwrap())?;
    let mut file = File::create(output_path)?;
    file.write_all(data)?;
    Ok(())
}

// Test 3: Archive extraction with canonicalize but insufficient validation
fn extract_archive(entry_name: &str, base_dir: &Path, content: &[u8]) -> std::io::Result<()> {
    let full_path = base_dir.join(entry_name);
    // VULNERABLE: canonicalize may not prevent all traversal attacks
    std::fs::create_dir_all(full_path.parent().unwrap())?;
    let mut file = File::create(full_path)?;
    file.write_all(content)?;
    Ok(())
}

// Test 4: Using user-provided filename from archive
fn save_archive_entry(filename: &str, content: &[u8], upload_dir: &Path) -> std::io::Result<()> {
    // VULNERABLE: filename from archive could contain ../
    let file_path = upload_dir.join(filename);
    let mut file = File::create(file_path)?;
    file.write_all(content)?;
    Ok(())
}

// Test 5: Zip crate usage without validation
fn unpack_zip_entry(entry_path: &str, destination: &Path, data: Vec<u8>) -> std::io::Result<()> {
    let target_path = destination.join(entry_path);
    // VULNERABLE: No check if target_path escapes destination
    if let Some(parent) = target_path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(target_path, data)?;
    Ok(())
}
